### PR TITLE
Added CUDA 11.0.2 and related recipes, incl. gompic/2020a and iccifortcuda/2020a

### PR DESCRIPTION
**Pull request by bartoldeman**
Like #10935 this introduces a new CUDAcore easyconfig to share
CUDA between use of GCC and Intel compilers, but
unlike #10935 this uses a versionsuffix for UCX+CUDA so it does
not need any framework changes or MODULEPATH adjustments.

### CODE REVIEW

**Comment by bartoldeman** [comment]
@boegelbot please test @ generoso

**Comment by Micket** [comment]
First; I agree with this approach compared to the other alternatives presented, though this got me thinking about the possibilities here:
If the MPI libraries only cares about whether or not the UCX library has CUDA support, this would let us avoid the `foss`+ `fosscuda` split. CUDA enabled things would simply live under `foss` and just depend on `('UCX', '1.8.0', '-CUDA-11.0.2')` (presumably adding the same versionsuffix themselves), i.e. something like this:
```
CUDA-11.0.2.eb # at SYSTEM level
UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb  # depends on CUDA-11.0.2
HPL-2.3-foss-2019b-CUDA-11.0.2.eb # depends on UCX/1.8.0-CUDA-11.0.2
```

A relatively minor thing, but, if we ever want to do something like that, it sure would be nicer if the system-level CUDA got to keep using the `CUDA` name, and we used some other name `CUDA.GCC` or whatever for the module under the compiler level. 
Simply adding an extra line to:
https://github.com/easybuilders/easybuild-framework/blob/7e1ec1c2903dfbe8afd27ca8e140ac58302e581c/easybuild/tools/module_naming_scheme/hierarchical_mns.py#L53-L69
should solve the HMNS issue (just like was done for the icc+ifort merge), and flat naming schemes don't care either way.
(though, it we were to actually opt for combining fosscuda into foss, then we wouldn't need this change, as it simply would stop existing)

*While writing this, I realized there might be some issues for those who RPATH the UCX libs into their openmpi, installation, which would require multiple versions of openmpi, forever forcing the use of separate `foss` and `fosscuda`. :frowning_face: )*

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
**FAILED**
Build succeeded for 1 out of 9 (9 easyconfigs in this PR)
build-node.computecanada.ca - Linux centos linux 7.8.2003, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 2.7.13
See https://gist.github.com/c664265f16331164f018c2bf43a1cdbe for a full test report.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
**FAILED**
Build succeeded for 8 out of 9 (9 easyconfigs in this PR)
build-node.computecanada.ca - Linux centos linux 7.8.2003, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.7.7
See https://gist.github.com/5a5b6e655e4d6e29a774302b76beaa50 for a full test report.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
**SUCCESS**
Build succeeded for 9 out of 9 (9 easyconfigs in this PR)
build-node.computecanada.ca - Linux centos linux 7.8.2003, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.7.7
See https://gist.github.com/10cde85212b91406ae7f7bada4823d96 for a full test report.

**Comment by bartoldeman** [comment]
@Micket it's not clear to me from your comment what I should change to this commit -- you said you were ok but did not approve it.

File: easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb
Comment (boegel) [inline easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb:6]: Please update to 1.9.0 (released just days ago)?

File: easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb
Comment (boegel) [inline easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: update to `11.1.0_455.23.05`?

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb
Comment (bartoldeman) [reply easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb:6]: I'd rather not do that for consistency with the rest of the 2020a components.

Comment (boegel) [reply easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb:6]: Actually, we should stick to the same version we use as a dep for OpenMPI in `gompi/2020a`, so leave this as is.

File: easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb
Comment (boegel) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: @bartoldeman bumped to CUDA 11.1.0 in https://github.com/ComputeCanada/easybuild-easyconfigs/pull/13

Comment (mboisson) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: No ?

Comment (mboisson) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: As commented on the PR on your PR, we had already settled on 11.0.2 for 2020a... why change now ? This won't work as we have already software installed with 2020a = 11.0.2 for CUDA.

Comment (mboisson) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: Also, bumping it to 11.1.0 basically means it can't be used on any of our clusters yet, since the driver requirements for 11.0 just got met by our clusters.

Comment (boegel) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: @mboisson I understand that this is a concern for ComputeCanada, but in the central repo we haven't settled on a CUDA version for `2020a` yet...

Are there other reasons to stick to an older CUDA version?

@akesandgren, @Micket, @lexming Thoughts on this?

Comment (mboisson) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: 2020a should use tools that were released around 2020a (i.e. first half of 2020) not that were released in the late 2020... if this PR had extended until 2022, would we release 2020a with CUDA 13.0 ? I hope not...

Comment (boegel) [reply easybuild/easyconfigs/c/CUDAcore/CUDAcore-11.0.2.eb:4]: I discussed this a bit with @mboisson, and I can see why motivating to go with CUDA 11.1 now for the `2020a` generation is difficult to motivate, given the significant delay in settling on `fosscuda/2020a`.

Let's keep CUDA 11.1 for `fosscuda/2020b`, since it is compatible with GCC 10 according to the [release notes](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html)

**Comment by boegel** [comment]
To avoid the `foss`/`fosscuda` split, I think we should add support for "optional dependencies", where EasyBuild can be configured to include CUDA as a dep or not.

On GPU systems you would then configure EasyBuild with `--with-optional-deps=CUDA` or something, which would active the optional CUDA dependency in UCX.

That's a good way forward to collapse `foss` and `fosscuda` I think...

**Comment by mboisson** [comment]
Probably a good idea, but beyond this PR :P

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 9 out of 9 (9 easyconfigs in this PR)
node3301.joltik.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6242 CPU @ 2.80GHz (cascadelake), Python 3.6.8
See https://gist.github.com/5edfd5e3b69a624567046276f0f19c8c for a full test report.

**Comment by boegel** [comment]
> Probably a good idea, but beyond this PR :P

Yeah, I would like to go for an approach like that in the upcoming `2020b` generation...

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 9 out of 9 (9 easyconfigs in this PR)
node2706.swalot.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 2.7.5
See https://gist.github.com/79b212dd5b48add3fb9f1b85c4549fb6 for a full test report.

**Review by boegel — DISMISSED** [review]
I think this is ready to go...

@Micket @lexming @akesandgren Any last words on this before we merge it?

File: easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb
Comment (akesandgren) [inline easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb:37]: This should use the predefined variable

**Review by akesandgren — CHANGES_REQUESTED** [review]

**Comment by lexming** [comment]
I'll test this over the weekend

File: easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb
Comment (boegel) [reply easybuild/easyconfigs/u/UCX/UCX-1.8.0-GCCcore-9.3.0-CUDA-11.0.2.eb:37]: see https://github.com/ComputeCanada/easybuild-easyconfigs/pull/14

**Comment by boegel** [comment]
> I'll test this over the weekend

@lexming Any results yet?

**Review by lexming — COMMENTED** [review]
I would like to suggest an enhancement of this PR. UCX is missing support for GDRCopy and GPUDirect RDMA. Hence, the only CUDA transport available is `cuda_copy`. Please check my PR https://github.com/ComputeCanada/easybuild-easyconfigs/pull/15 to add a dependency on GDRCopy and enable it in UCX, this enables `gdr_copy` as an additional `UCX_TLS`. This seems to be a more standard installation. More info at https://github.com/openucx/ucx/wiki/NVIDIA-GPU-Support

**Comment by akesandgren** [comment]
+1 on GDRCopy and GPUDirect, it was something I never got around to fiddle with in our UCX-CUDA build...

**Comment by lexming** [comment]
I tested this build of UCX with CUDA in a non-GPU node and it works, but there is a big fat warning on every execution

```
$ mpirun ./test
[node101.hydra.os:17605] [[48239,0],0] ORTE_ERROR_LOG: Out of resource in file util/show_help.c at line 501
--------------------------------------------------------------------------
The library attempted to open the following supporting CUDA libraries,
but each of them failed.  CUDA-aware support is disabled.
libcuda.so.1: cannot open shared object file: No such file or directory
libcuda.dylib: cannot open shared object file: No such file or directory
/usr/lib64/libcuda.so.1: cannot open shared object file: No such file or directory
/usr/lib64/libcuda.dylib: cannot open shared object file: No such file or directory
If you are not interested in CUDA-aware support, then run with
--mca opal_warn_on_missing_libcuda 0 to suppress this message.  If you are interested
in CUDA-aware support, then try setting LD_LIBRARY_PATH to the location
of libcuda.so.1 to get passed this issue.
--------------------------------------------------------------------------
Hello, I am rank 0/2
Hello, I am rank 1/2
```

This is absolutely fine for this PR, but we should have this in mind in our quest for the one UCX to rule them all. Easiest solution would be to just disable the warning with the envar `OMPI_MCA_opal_warn_on_missing_libcuda=0`

**Comment by bartoldeman** [comment]
@boegelbot please test @ generoso

**Comment by bartoldeman** [comment]
Fails on generoso because of
```
ldconfig -n /tmp/boegelbot/GDRCopy/2.1/GCCcore-9.3.0-CUDA-11.0.2/gdrcopy-2.1/src
make[1]: ldconfig: Command not found
```
I'll check what can be done...

**Comment by bartoldeman** [comment]
I'll try doing something like prebuildopts="PATH=$PATH:/sbin && " because ldconfig usually sits there and sbin is not always in the path -- but need to be offline for a while.

**Comment by boegel** [comment]
> I'll try doing something like prebuildopts="PATH=$PATH:/sbin && " because ldconfig usually sits there and sbin is not always in the path -- but need to be offline for a while.

`ldconfig` is indeed at `/usr/sbin/ldconfig` on generoso

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
node3408.kirlia.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 2.7.5
See https://gist.github.com/9cce610aa7571502d5ae5f1c8595cbb0 for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
node154.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/076daad50492f360566c6603017539da for a full test report.

**Review by lexming — DISMISSED** [review]
LGTM

**Comment by bartoldeman** [comment]
@boegelbot please test @ generoso

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
build-node.computecanada.ca - Linux centos linux 7.8.2003, x86_64, Intel Xeon Processor (Skylake, IBRS), Python 3.7.7
See https://gist.github.com/0d91a9465733179f6fda3926dda1714f for a full test report.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
node154.hydra.os - Linux centos linux 7.7.1908, x86_64, Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz, Python 2.7.5
See https://gist.github.com/b939306feb9a8364a30104fef72f4706 for a full test report.

**Review by lexming — APPROVED** [review]
LGTM

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
node3404.kirlia.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz (cascadelake), Python 2.7.5
See https://gist.github.com/6e437eed7ddba02daa96a7badf2e04ec for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 12 out of 12 (11 easyconfigs in this PR)
node3502.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7302P 16-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/f45264ebb3d1328c78a0b79dd6e4f509 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
node2424.golett.os - Linux centos linux 7.8.2003, x86_64, Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (haswell), Python 2.7.5
See https://gist.github.com/97c627d65aa9b3895e55449e50e26eaa for a full test report.

**Comment by zao** [comment]
Test report by @zao
**SUCCESS**
Build succeeded for 11 out of 11 (11 easyconfigs in this PR)
freja - Linux Ubuntu 20.04, x86_64, Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz, Python 3.8.2
See https://gist.github.com/c9fdbe7ed70ed6a0f989b6f12a48bd6f for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 8 out of 8 (8 easyconfigs in this PR)
bear-pg0306u19a.bear.cluster - Linux RHEL 8.2, POWER, 8335-GTX (power9le), Python 3.6.8
See https://gist.github.com/7ddd170bfa042bbffd88aa27b610ca89 for a full test report.

**Comment by boegel** [comment]
Going in, thanks @bartoldeman!

**Pull request merged**
